File: sdkcode/php/UniTranslate.php
<?php

class UniTranslate {
    /**
     * Generate HMAC-SHA256 signature for the given parameters
     * @param string $key Secret key
     * @param array $params Parameters to sign
     * @return string Hex-encoded signature
     */
    public static function authEncrypt(string $key, array $params): string {
        $data = self::sortMapToStr($params);
        return hash_hmac('sha256', $data, $key);
    }

    /**
     * Convert map to sorted string representation
     * @param array $data Input data
     * @return string Sorted string representation
     */
    private static function sortMapToStr(array $data): string {
        if (empty($data)) {
            return '';
        }

        ksort($data);
        $parts = [];

        foreach ($data as $k => $v) {
            if (is_array($v)) {
                if (self::isAssoc($v)) {
                    // Handle nested map
                    $value = '{' . self::sortMapToStr($v) . '}';
                } else {
                    // Handle array
                    $arrayParts = [];
                    foreach ($v as $item) {
                        if (is_array($item) && self::isAssoc($item)) {
                            $arrayParts[] = '{' . self::sortMapToStr($item) . '}';
                        } else {
                            $arrayParts[] = is_null($item) ? '' : (string)$item;
                        }
                    }
                    $value = '[' . implode(',', $arrayParts) . ']';
                }
            } else {
                $value = is_null($v) ? '' : (string)$v;
            }
            $parts[] = $k . ':' . $value;
        }

        return implode('&', $parts);
    }

    /**
     * Check if array is associative
     * @param array $arr Array to check
     * @return bool True if associative
     */
    private static function isAssoc(array $arr): bool {
        if (empty($arr)) return false;
        return array_keys($arr) !== range(0, count($arr) - 1);
    }
}

File: sdkcode/python/uni_translate.py
import hmac
import hashlib
from typing import Dict, Any, Union

class UniTranslate:
    @staticmethod
    def auth_encrypt(key: str, params: Dict[str, Any]) -> str:
        """
        Generate HMAC-SHA256 signature for the given parameters
        Args:
            key: Secret key
            params: Parameters to sign
        Returns:
            Hex-encoded signature
        """
        data = UniTranslate.sort_map_to_str(params)
        hmac_obj = hmac.new(
            key.encode('utf-8'),
            data.encode('utf-8'),
            hashlib.sha256
        )
        return hmac_obj.hexdigest()

    @staticmethod
    def sort_map_to_str(data: Dict[str, Any]) -> str:
        """
        Convert map to sorted string representation
        Args:
            data: Input data
        Returns:
            Sorted string representation
        """
        if not data:
            return ''

        parts = []
        for k in sorted(data.keys()):
            v = data[k]
            if isinstance(v, dict):
                # Handle nested map
                value = '{' + UniTranslate.sort_map_to_str(v) + '}'
            elif isinstance(v, (list, tuple)):
                # Handle array
                array_parts = []
                for item in v:
                    if isinstance(item, dict):
                        array_parts.append('{' + UniTranslate.sort_map_to_str(item) + '}')
                    else:
                        array_parts.append(str(item if item is not None else ''))
                value = '[' + ','.join(array_parts) + ']'
            else:
                value = str(v if v is not None else '')
            parts.append(f"{k}:{value}")

        return '&'.join(parts)
